SG14/inplace_function.h

```diff
@@ -150,6 +150,76 @@ struct is_valid_inplace_dst : std::true_type
     );
 };
 
+// C++11 MSVC compatible implementation of std::is_invocable_r,
+// std version doesn't seem to respect return types and requires C++17.
+// See https://godbolt.org/z/T0bd3K for defect example.
+
+template<
+    typename Default,
+    template<typename...> class Expression,
+    typename... Ts
+>
+class detected_or
+{
+private:
+    template<
+        template<typename...> class Alias,
+        typename... Args,
+        typename Ret_t = Alias<Args...>
+    >
+    static auto ov(int ph) -> Ret_t;
+
+    template<
+        template<typename...> class Alias,
+        typename... Args
+    >
+    static Default ov(...);
+
+public:
+    using type = decltype(ov<Expression, Ts...>(0));
+};
+
+template<
+    typename R,
+    typename C,
+    typename... Args
+>
+using def_closure_return_type_convertible = std::is_convertible<
+    typename std::decay<R>::type,
+    decltype(std::declval<C>()(std::declval<Args>()...))
+>;
+
+template<
+    typename R,
+    typename C,
+    typename... Args
+>
+using is_invocable_r = typename detected_or<
+    std::false_type,
+    def_closure_return_type_convertible,
+    R, C, Args...
+>::type;
+
+static inline int return_int() { return 3; }
+static inline int return_int_with_parameter(int) { return 5; };
+static inline void return_void_with_ref_parameter(int&) {}
+
+using C_Int1 = typename std::decay<decltype(return_int)>::type;
+using C_Int2 = typename std::decay<decltype(return_int_with_parameter)>::type;
+using C_Void = typename std::decay<decltype(return_void_with_ref_parameter)>::type;
+
+static_assert(is_invocable_r<int, C_Int1>::value, "IV A");
+static_assert(! is_invocable_r<int, C_Int2>::value, "IV B");
+static_assert(! is_invocable_r<int, C_Void>::value, "IV C");
+
+static_assert(is_invocable_r<int, C_Int2, int>::value, "IV D");
+static_assert(! is_invocable_r<int, C_Int1, int>::value, "IV E");
+static_assert(! is_invocable_r<int, C_Void, int>::value, "IV F");
+
+static_assert(is_invocable_r<void, C_Void, int&>::value, "IV G");
+static_assert(! is_invocable_r<void, C_Int1, int&>::value, "IV H");
+static_assert(! is_invocable_r<void, C_Int2, int&>::value, "IV I");
```

Comment:
I would be inclined to move all these test cases out of the header file and into the unit-tests. I know I set a bad example already with `aligned_storage`, though. 😛 

If they're kept here, at least I think they should be simplified down to just in-line function types, i.e.

    static_assert(is_invocable_r<int, int()>::value, "IV A");
    static_assert(!is_invocable_r<int, int(int)>::value, "IV B");
    static_assert(!is_invocable_r<int, void(int&)>::value, "IV C");

That's easier to read, and doesn't introduce new entities into the user's program, and I don't think it loses anything important in terms of coverage.

Reply:
My intuition was to put a test in test file, yet I didn't want to break with the precedent `aligned_storage` set so I put them there, I'll gladly move them to the test file.